nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -645,8 +645,85 @@ impl StringPrototype {
+    fn replace(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? RequireObjectCoercible(this value).
+        let o = require_object_coercible(agent, this_value)?;
+
+        let search_value = args.get(0);
+        let replace_value = args.get(1);
+
+        // 2. If searchValue is neither undefined nor null, then
+        if !search_value.is_null() && !search_value.is_undefined() {
+            let symbol = WellKnownSymbolIndexes::Replace.into();
+            // If replacer is not undefined, Return ? Call(replacer, searchValue, « O, replaceValue »).
+            if let Ok(Some(replacer)) = get_method(agent, search_value, symbol) {
+                return call_function(
+                    agent,
+                    replacer,
+                    search_value,
+                    Some(ArgumentsList(&[o, replace_value])),
+                );
+            }
+        }
+
+        // 3. Let s be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 4. Let searchString be ? ToString(searchValue).
+        let search_string = to_string(agent, search_value)?;
+
+        // 5. Let functionalReplace be IsCallable(replaceValue).
+        let functional_replace = is_callable(replace_value);
+
+        // 6. If functionalReplace is false, Set replaceValue to ? ToString(replaceValue).
+        let replace_value = match functional_replace {
+            None => Some(to_string(agent, replace_value)?),
+            Some(_) => None,
+        };
+
+        // 7. Let searchLength be the length of searchString.
+        let search_length = search_string.len(agent);
+
+        // 8. Let position be StringIndexOf(s, searchString, 0).
+        // 9. If position is not-found, return s.
+        let position = if let Some(position) = search_string.as_str(agent).find(s.as_str(agent)) {
+            position
+        } else {
+            return Ok(s.into_value());
+        };
+
+        // Use rust's replace if the replace is a string, not a function
+        if let Some(replace_value) = replace_value {
+            let result = s
+                .as_str(agent)
+                .replace(search_string.as_str(agent), replace_value.as_str(agent));
+            return Ok(String::from_string(agent, result).into_value());
+        }
+
+        // 10. Let preceding be the substring of s from 0 to position.
+        // 11. Let following be the substring of s from position + searchLength.
+        // 12. If functionalReplace is true,
+        let preceding = &s.as_str(agent)[0..position].to_owned();
+        let following = &s.as_str(agent)[position..search_length].to_owned();
+
+        // Let replacement be ? ToString(? Call(replaceValue, undefined, « searchString, 𝔽(position), string »)).
+        let result = call(
+            agent,
+            functional_replace.unwrap().into_value(),
```

Comment:
**note:** This unwrap here kind of points to this function probably benefiting from a bit of a refactor: It would be nice if the functional_replace and/or replace_value path could be done entirely within a `match` branch or perhaps a `if let Some(..)/Ok(..) = .. {}` branch.